linera-chain/src/chain.rs

```diff
@@ -794,197 +743,118 @@ where
-                    } else {
-                        Vec::new()
+                Transaction::ExecuteOperation(operation) => {
+                    #[cfg(with_metrics)]
+                    let _operation_latency = OPERATION_EXECUTION_LATENCY.measure_latency();
+                    let context = OperationContext {
+                        chain_id,
+                        height: block.height,
+                        index: Some(txn_index),
+                        authenticated_signer: block.authenticated_signer,
+                        authenticated_caller_id: None,
                     };
-                    (outcomes, Vec::new())
+                    self.execution_state
+                        .execute_operation(
+                            context,
+                            local_time,
+                            operation.clone(),
+                            &mut txn_tracker,
+                            &mut resource_controller,
+                        )
+                        .await
+                        .map_err(with_context)?;
                 }
-            };
-            new_oracle_responses.push(oracle_responses);
-            let (messages_out, new_events) = self
-                .process_execution_outcomes(context.height, outcomes)
+            }
+
+            self.execution_state
+                .update_execution_outcomes_with_app_registrations(&mut txn_tracker)
+                .await
+                .map_err(with_context)?;
+            let (txn_outcomes, txn_oracle_responses, new_next_message_index) =
+                txn_tracker.destructure().map_err(with_context)?;
+            next_message_index = new_next_message_index;
+            let (txn_messages, txn_events) = self
+                .process_execution_outcomes(block.height, txn_outcomes)
                 .await?;
-            if let MessageAction::Accept = message.action {
-                for message_out in &messages_out {
+            match transaction {
+                Transaction::ExecuteOperation(operation) => {
                     resource_controller
                         .with_state(&mut self.execution_state)
                         .await?
-                        .track_message(&message_out.message)
-                        .map_err(|err| ChainError::ExecutionError(err, chain_execution_context))?;
+                        .track_operation(operation)
+                        .map_err(with_context)?;
+                }
+                Transaction::ReceiveMessages(incoming_bundle) => {
+                    if let MessageAction::Accept = incoming_bundle.action {
+                        for message_out in &txn_messages {
+                            resource_controller
+                                .with_state(&mut self.execution_state)
+                                .await?
+                                .track_message(&message_out.message)
+                                .map_err(with_context)?;
+                        }
```

Comment:
Is this executed twice when receiving messages? There's the same code right after the `match` expression.

Reply:
Good catch, thanks! It looks like we tracked the outgoing messages from handling _accepted messages_ or _operations_ (i.e. everything except _rejected_ messages).